Earthfile

```diff
@@ -0,0 +1,399 @@
+      # https://github.com/earthly/earthly/issues/4143 is fixed.
+      RUN gotestsum --no-color=false --format testname --junitfile e2e-tests.xml --raw-command go tool test2json -t -p E2E ./e2e -test.v ${FLAGS}
+    FINALLY
+      SAVE ARTIFACT --if-exists e2e-tests.xml AS LOCAL _output/tests/e2e-tests.xml
+    END
+  END
+
+# hack builds Crossplane, and deploys it to a kind cluster. It runs in your
+# local environment, not a container. The kind cluster will keep running until
+# you run the unhack target. Run hack again to rebuild Crossplane and restart
+# the kind cluster with the new build.
+hack:
+  # TODO(negz): This could run an interactive shell inside a temporary container
+  # once https://github.com/earthly/earthly/issues/3206 is fixed.
+  LOCALLY
+  WAIT
+    BUILD +unhack
+  END
+  COPY +helm-setup/helm .hack/helm
+  COPY +kind-setup/kind .hack/kind
+  COPY (+helm-build/output --CROSSPLANE_VERSION=v0.0.0-hack) .hack/charts
+  WITH DOCKER --load crossplane-hack/crossplane:hack=+image
+    RUN \
+      .hack/kind create cluster --name crossplane-hack && \
+      .hack/kind load docker-image --name crossplane-hack crossplane-hack/crossplane:hack && \
+      .hack/helm install --create-namespace --namespace crossplane-system crossplane .hack/charts/crossplane-0.0.0-hack.tgz \
+        --set "image.pullPolicy=Never,image.repository=crossplane-hack/crossplane,image.tag=hack" \
+        --set "args={--debug}"
+  END
+  RUN docker image rm crossplane-hack/crossplane:hack
+  RUN rm -rf .hack
+
+# unhack deletes the kind cluster created by the hack target.
+unhack:
+  LOCALLY
+  COPY +kind-setup/kind .hack/kind
+  RUN .hack/kind delete cluster --name crossplane-hack
+  RUN rm -rf .hack
+
+# go-modules downloads Crossplane's go modules. It's the base target of most Go
+# related target (go-build, etc).
+go-modules:
+  ARG NATIVEPLATFORM
+  FROM --platform=${NATIVEPLATFORM} golang:${GO_VERSION}
+  WORKDIR /crossplane
+  CACHE --id go-build --sharing shared /root/.cache/go-build
+  COPY go.mod go.sum ./
+  RUN go mod download
+  SAVE ARTIFACT go.mod AS LOCAL go.mod
+  SAVE ARTIFACT go.sum AS LOCAL go.sum
+
+# go-modules-tidy tidies and verifies go.mod and go.sum.
+go-modules-tidy:
+  FROM +go-modules
+  CACHE --id go-build --sharing shared /root/.cache/go-build
+  COPY --dir apis/ cmd/ internal/ pkg/ test/ .
+  RUN go mod tidy
+  RUN go mod verify
+  SAVE ARTIFACT go.mod AS LOCAL go.mod
+  SAVE ARTIFACT go.sum AS LOCAL go.sum
+
+# go-generate runs Go code generation.
+go-generate:
+  FROM +go-modules
+  CACHE --id go-build --sharing shared /root/.cache/go-build
+  COPY --dir hack/ apis/ internal/ .
+  RUN go generate -tags 'generate' ./apis/...
+  SAVE ARTIFACT apis/ AS LOCAL apis
+
+# go-build builds Crossplane binaries for your native OS and architecture.
+go-build:
+  ARG EARTHLY_GIT_SHORT_HASH
+  ARG EARTHLY_GIT_COMMIT_TIMESTAMP
+  ARG CROSSPLANE_VERSION=v0.0.0-${EARTHLY_GIT_COMMIT_TIMESTAMP}-${EARTHLY_GIT_SHORT_HASH}
```

Comment:
I think with the `make` based approach, we would get a higher fidelity version number when building locally, e.g. one that is based on `git tags` and looks like ` v1.16.0-rc.0.268.gf66104f1.dirty`. I liked having that since it was informative for what upstream crossplane version are my local changes based on.

Is it possible to keep that behavior?

Reply:
I wouldn't recommend it.

AFACIT it's not possible to generate versions like this without access to a full, deep copy of the entire git repo. In a containerized build system like Earthly (or Docker) that means running `COPY . .` to copy the entire repo into the image. Doing that is pretty slow, but more importantly it means _any_ change in the repo will invalidate the layer cache and trigger a full rebuild.

https://docs.earthly.dev/docs/guides/best-practices#copy-only-the-minimal-amount-of-files.-avoid-copying-.git

FWIW I'm increasingly not a fan of trying to extract version information from git tags. I moved away from that pattern with functions - there you just specify what version you want at build time. I didn't want to go that far with this change though, which is why there's a hacky `ci-version` target that tries to maintain the behavior for CI builds.